### PR TITLE
Fix (non-local IP address) ZAP API loop

### LINKED ISSUE #2318: ZAP Error [java.net.SocketTimeoutException]: Read timed out when running on AWS EC2 instance
When I ssh to EC2 instance and run:

`./zap.sh -daemon -port 8090 -host 0.0.0.0`

ZAP starts without any problem:

> Found Java version 1.7.0_95
> Available memory: 16047 MB
> Setting jvm heap size: -Xmx512m
> 0 [main] INFO org.zaproxy.zap.DaemonBootstrap  - OWASP ZAP 2.4.3 started.
> 325 [main] INFO hsqldb.db.HSQLDB379AF3DEBD.ENGINE  - dataFileCache open start
> 332 [main] INFO hsqldb.db.HSQLDB379AF3DEBD.ENGINE  - dataFileCache open end
> 657 [main] INFO org.parosproxy.paros.network.SSLConnector  - Reading supported SSL/TLS protocols...
> 657 [main] INFO org.parosproxy.paros.network.SSLConnector  - Using a SSLEngine...
> 872 [main] INFO org.parosproxy.paros.network.SSLConnector  - Done reading supported SSL/TLS protocols: [SSLv2Hello, SSLv3, TLSv1, TLSv1.1, TLSv1.2]
> 876 [main] INFO org.parosproxy.paros.extension.option.OptionsParamCertificate  - Unsafe SSL renegotiation disabled.
> 886 [ZAP-daemon] INFO org.zaproxy.zap.control.ExtensionFactory  - Loading extensions
> 1583 [ZAP-daemon] INFO org.zaproxy.zap.control.ExtensionFactory  - Extensions loaded
> 1888 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Change user agent to other browsers. 
> 1888 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Detect insecure or potentially malicious content in HTTP responses.
> 1888 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Detect and alert 'Set-cookie' attempt in HTTP response for modification.
> 1888 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Avoid browser cache (strip off IfModifiedSince)
> 1888 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Log cookies sent by browser.
> 1888 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Log unique GET queries into file:filter/get.xls
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Log unique POST queries into file:  filter/post.xls
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Log request and response into file: filter/message.txt
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Replace HTTP request body using defined pattern.
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Replace HTTP request header using defined pattern.
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Replace HTTP response body using defined pattern.
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Replace HTTP response header using defined pattern.
> 1889 [ZAP-daemon] INFO org.parosproxy.paros.extension.filter.FilterFactory  - loaded filter Send ZAP session request ID
> 2016 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Allows ZAP to check for updates
> 2019 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionViewOption
> 2019 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionEdit
> 2019 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionFilter
> 2019 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Provides a rest based API for controlling and accessing ZAP
> 2126 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionState
> 2126 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHistory
> 2127 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Show hidden fields and enable disabled fields
> 2128 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Search messages for strings and regular expressions 
> 2129 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Encode/Decode/Hash...
> 2129 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Allows you to intercept and modify requests and responses
> 2129 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Passive scanner
> 2170 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Script passive scan rules
> 2170 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Application Error Disclosure
> 2170 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Incomplete or No Cache-control and Pragma HTTP Header Set
> 2170 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Content-Type Header Missing
> 2170 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Cookie No HttpOnly Flag
> 2171 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Cookie Without Secure Flag
> 2171 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Cross-Domain JavaScript Source File Inclusion
> 2171 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Web Browser XSS Protection Not Enabled
> 2171 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Secure Pages Include Mixed Content
> 2171 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Password Autocomplete in Browser
> 2172 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Private IP Disclosure
> 2172 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: Session ID in URL Rewrite
> 2172 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: X-Content-Type-Options Header Missing
> 2172 [ZAP-daemon] INFO org.zaproxy.zap.extension.pscan.ExtensionPassiveScan  - loaded passive scan rule: X-Frame-Options Header Not Set
> 2191 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Allows you to view and manage alerts
> 2191 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Active scanner, heavily based on the original Paros active scanner, but with additional tests added
> 2196 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Spider used for automatically finding URIs on a site
> 2202 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing A set of common popup menus for miscellaneous tasks
> 2202 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Forced browsing of files and directories using code from the OWASP DirBuster tool
> 2203 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionManualRequest
> 2203 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Compares 2 sessions and generates an HTML file showing the differences
> 2203 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Invoke external applications passing context related information such as URLs and parameters
> 2204 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Handles anti cross site request forgery (CSRF) tokens
> 2206 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionAuthentication
> 2809 [ZAP-daemon] INFO org.zaproxy.zap.extension.authentication.ExtensionAuthentication  - Loaded authentication method types: [Form-based Authentication, HTTP/NTLM Authentication, Manual Authentication, Script-based Authentication]
> 2810 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Creates a dynamic SSL certificate to allow SSL communications to be intercepted without warnings being generated by the browser
> 2811 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Logs errors to the Output tab in development mode only
> 2811 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionUserManagement
> 2812 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Summarise and analyse FORM and URL parameters as well as cookies
> 2812 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Script integration
> 2820 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Scripting console, supports all JSR 223 scripting languages
> 2820 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionForcedUser
> 2821 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Extension handling HTTP sessions
> 2821 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Zest is a specialized scripting language from Mozilla specifically designed to be used in security tools
> 2887 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionDiff
> 2888 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionRequestPostTableView
> 2888 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Simple browser configuration
> 2888 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionSessionManagement
> 2983 [ZAP-daemon] INFO org.zaproxy.zap.extension.sessions.ExtensionSessionManagement  - Loaded session management method types: [Cookie-based Session Management, Http Authentication Session Management]
> 2983 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelRequestFormTableView
> 2983 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Capture messages from WebSockets with the ability to set breakpoints.
> 2986 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Summarise and analyse FORM and URL parameters as well as cookies
> 2986 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionAuthorization
> 2987 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing AJAX Spider, uses Crawljax
> 2987 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Handles adding Global Excluded URLs
> 2988 [ZAP-daemon] WARN org.zaproxy.zap.extension.globalexcludeurl.ExtensionGlobalExcludeURL  - GlobalExcludeURL.optionsLoaded()
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Adds menu item to refresh the Sites tree
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Adds support for configurable keyboard shortcuts for all of the ZAP menus.
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing OWASP ZAP User Guide
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionReport
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Allows you to configure which extensions are loaded when ZAP starts 
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelComponentonentAll
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelHexView
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelImageView
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelLargeRequestView
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelLargeResponseView
> 2988 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelRequestQueryCookieTableView
> 2989 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionHttpPanelSyntaxHighlightTextView
> 2989 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Active Scan Rules
> 2989 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing The ZAP Getting Started Guide
> 2989 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Tips and Tricks
> 2989 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Provides the foundation for concrete message types (for example, HTTP, WebSockets) expose fuzzer implementations.
> 2990 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Allows to fuzz HTTP messages.
> 2991 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing The Online menu links
> 2991 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing ExtensionSaveRawHttpMessage
> 2991 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Provides WebDrivers to control several browsers using Selenium and includes HtmlUnit browser.
> 2991 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Translations of the core language files
> 2991 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Passive Scan Rules
> 2991 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Quick Start panel 
> 2992 [ZAP-daemon] INFO org.parosproxy.paros.extension.ExtensionLoader  - Initializing Allows to fuzz WebSocket messages.
> 2993 [Thread-4] INFO org.zaproxy.zap.extension.dynssl.ExtensionDynSSL  - Creating new root CA certificate
> 3089 [ZAP-daemon] INFO org.zaproxy.zap.DaemonBootstrap  - ZAP is now listening on 0.0.0.0:8090
> 3446 [Thread-4] INFO org.zaproxy.zap.extension.dynssl.ExtensionDynSSL  - New root CA certificate created

I ensure that ZAP listens to port 8090 by running netstat -uta | grep 8090:

`tcp6       0      0 [::]:8090               [::]:*                  LISTEN`

Then I use curl from my laptop to access EC2 instance by it's public IP:

> curl -v EC2-host:8090
> - Rebuilt URL to: EC2-host:8090/
> -   Trying EC2-host...
> - Connected to EC2-host (EC2-host) port 8090 (#0)
>   GET / HTTP/1.1
>   Host: EC2-host:8090
>   User-Agent: Mozilla/5.0 (compatible; MSIE 9.0; Windows NT 6.1; Trident/5.0)
>   Accept: _/_
>   Referer: 
> 
> < HTTP/1.1 504 Gateway Timeout
> < Content-Length: 1911
> < Content-Type: text/plain; charset=UTF-8
> - Connection #0 to host 52.37.251.50 left intact

at the same time netstat shows me a lot of opened connections:

> tcp        0      0 ip-172-30-0-221:8090    EC2-host.:34088 SYN_RECV  
> tcp6       0      0 [::]:8090               [::]:\*                  LISTEN  
> tcp6       0      0 ip-172-30-0-221:38852   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:55390   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:38020 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:52432   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:55759   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:58787 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:52077 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:49051 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:55484 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:40724 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:47354   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:36019   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:52022   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:51511 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:53242   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:35760   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:39354   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:59507   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:53388   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:36219 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:52108 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:34705 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:49159   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:48851   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:45609 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:45546   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:50533   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:36484   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:43129 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:38790   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:50698 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:33265   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:44329 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:51403 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:40302 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:44767 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:47802   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:8090    EC2-host.:47444 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:36401   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:39548   EC2-host.u:8090 ESTABLISHED
> tcp6       0      0 ip-172-30-0-221:56141   EC2-host.u:8090 ESTABLISHED
> ........

Eventually ZAP throws: 

> 217142 [ZAP-ProxyThread-45792] ERROR org.zaproxy.zap.ZAP$UncaughtExceptionLogger  - Exception in thread "ZAP-ProxyThread-45792"
> java.lang.OutOfMemoryError: GC overhead limit exceeded
> at java.net.SocketInputStream.socketRead0(Native Method)
> at java.net.SocketInputStream.read(SocketInputStream.java:152)
> at java.net.SocketInputStream.read(SocketInputStream.java:122)
> at java.io.BufferedInputStream.fill(BufferedInputStream.java:235)
> at java.io.BufferedInputStream.read(BufferedInputStream.java:254)
> at org.apache.commons.httpclient.HttpParser.readRawLine(HttpParser.java:78)
> at org.apache.commons.httpclient.HttpParser.readLine(HttpParser.java:106)
> at org.apache.commons.httpclient.HttpConnection.readLine(HttpConnection.java:1116)
> at org.apache.commons.httpclient.MultiThreadedHttpConnectionManager$HttpConnectionAdapter.readLine(MultiThreadedHttpConnectionManager.java:1413)
> at org.apache.commons.httpclient.HttpMethodBase.readStatusLine(Unknown Source)
> at org.zaproxy.zap.ZapGetMethod.readResponse(Unknown Source)
> at org.apache.commons.httpclient.HttpMethodBase.execute(Unknown Source)
> at org.apache.commons.httpclient.HttpMethodDirector.executeWithRetry(Unknown Source)
> at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(Unknown Source)
> at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:397)
> at org.parosproxy.paros.network.HttpSender.executeMethod(Unknown Source)
> at org.parosproxy.paros.network.HttpSender.runMethod(Unknown Source)
> at org.parosproxy.paros.network.HttpSender.send(Unknown Source)
> at org.parosproxy.paros.network.HttpSender.sendAuthenticated(Unknown Source)
> at org.parosproxy.paros.network.HttpSender.sendAndReceive(Unknown Source)
> at org.parosproxy.paros.network.HttpSender.sendAndReceive(Unknown Source)
> at org.parosproxy.paros.core.proxy.ProxyThread.processHttp(Unknown Source)
> at org.parosproxy.paros.core.proxy.ProxyThread.run(Unknown Source)
> at java.lang.Thread.run(Thread.java:745)

At the same time ZAP API is accessible locally:

> curl -v http://127.0.0.1:8090
> - Rebuilt URL to: http://127.0.0.1:8090/
> - Hostname was NOT found in DNS cache
> -   Trying 127.0.0.1...
> - Connected to 127.0.0.1 (127.0.0.1) port 8090 (#0)
>   GET / HTTP/1.1
>   User-Agent: curl/7.35.0
>   Host: 127.0.0.1:8090
>   Accept: _/_
> 
> < HTTP/1.1 200 OK
> < Pragma: no-cache
> < Cache-Control: no-cache
> < Access-Control-Allow-Methods: GET,POST,OPTIONS
> < Access-Control-Allow-Headers: ZAP-Header
> < X-Clacks-Overhead: GNU Terry Pratchett
> < Content-Length: 1084
> < Content-Type: text/html; charset=UTF-8
> < 
> <head>
> <title>ZAP API UI</title>
> </head>
> <body>
> <h1>Welcome to the OWASP Zed Attack Proxy (ZAP)</h1><p>ZAP is an easy to use integrated penetration testing tool for finding vulnerabilities in web applications.</p><p></p><p>Please be aware that you should only attack applications that you have been specifically been given permission to test.</p><h2>Proxy Configuration</h2><p>To use ZAP effectively it is recommended that you configure your browser to proxy via ZAP.</p><p></p><p>You can do that manually or by configuring your browser to use the generated <a href="/OTHER/core/other/proxy.pac">PAC file</a>.</p></html><h2>Links</h2><li><a href="/UI">Local API</a></li><li><a href="https://www.owasp.org/index.php/ZAP">ZAP Homepage</a></li><li><a href="https://github.com/zaproxy/zaproxy/wiki/Introduction">ZAP Wiki</a></li><li><a href="http://groups.google.com/group/zaproxy-users">ZAP User Group</a></li><li><a href="http://groups.google.com/group/zaproxy-develop">ZAP Developer Group</a></li><li><a href="https://github.com/zaproxy/zaproxy/issues">Report an issue</a></li></body>
> - Connection #0 to host 127.0.0.1 left intact

So it looks very strange and I have no idea why so many connections are opened.

More discussions can be found here: https://groups.google.com/forum/#!topic/zaproxy-develop/sZg37GBkF7w

### PR DESCRIPTION
**Pull request by thc202**
Check if the requested address belongs to any of the network interfaces
when checking if it is recursive, that is, a request to ZAP itself,
preventing a request loop when the IP address is non-local.

Add the check to method ProxyThread.isRecursive(HttpRequestHeader) and
refactor, to reduce code complexity, by extracting simpler methods.

Fix #2318 - ZAP Error [java.net.SocketTimeoutException]: Read timed out
when running on AWS EC2 instance

### CODE REVIEW

File: src/org/parosproxy/paros/core/proxy/ProxyThread.java [hunk 1]
```diff
@@ -609,6 +611,20 @@ private boolean notifyPersistentConnectionListener(HttpMessage httpMessage, Sock
 		return keepSocketOpen;
 	}
 	
+	/**
+	 * Tells whether or not the given {@code header} has a request to the (parent) proxy itself.
+	 * <p>
+	 * The request is to the proxy itself if the following conditions are met:
+	 * <ol>
+	 * <li>The requested port is the one that the proxy is bond to;</li>
```
Comment (kingthorin) [inline src/org/parosproxy/paros/core/proxy/ProxyThread.java]: "...bound to;"

File: src/org/parosproxy/paros/core/proxy/ProxyThread.java [hunk 2]
```diff
@@ -616,13 +632,8 @@ private boolean isRecursive(HttpRequestHeader header) {
...
                 }
-                InetAddress targetAddress = InetAddress.getByName(targetDomain);
-                if (parentServer.getProxyParam().isProxyIpAnyLocalAddress()) {
-                    if (targetAddress.isLoopbackAddress() || targetAddress.isSiteLocalAddress()
-                            || targetAddress.isAnyLocalAddress()) {
-                        return true;
-                    }
-                } else if (targetAddress.equals(inSocket.getLocalAddress())) {
+
+                if (isProxyAddress(InetAddress.getByName(targetDomain))) {
```
Comment (kingthorin) [inline src/org/parosproxy/paros/core/proxy/ProxyThread.java]: Will this work as expected when bound to 0.0.0.0 (all interfaces)?

**Comment by kingthorin** [comment]
Did you test that this works if 0.0.0.0 (all interfaces) is the bind address?

File: src/org/parosproxy/paros/core/proxy/ProxyThread.java (hunk 1 above)
Comment (thc202) [inline src/org/parosproxy/paros/core/proxy/ProxyThread.java]: Fixed.

**Comment by thc202** [comment]
Yes, that was one of the tests.

**Comment by kingthorin** [comment]
Great. Looks good.

**Pull request merged**